src/stream/write/mod.rs

```diff
@@ -41,22 +41,25 @@ pub struct Decoder<'a, W: Write> {
 ///
 /// [`auto_finish()`]: Encoder::auto_finish
 /// [`Encoder`]: Encoder
-pub struct AutoFinishEncoder<'a, W: Write> {
+pub struct AutoFinishEncoder<'a, W: Write, F = Box<dyn FnMut(io::Result<W>)>>
+where
+    F: FnMut(io::Result<W>) + 'static,
+{
     // We wrap this in an option to take it during drop.
     encoder: Option<Encoder<'a, W>>,
 
     // TODO: make this a FnOnce once it works in a Box
-    on_finish: Option<Box<dyn FnMut(io::Result<W>)>>,
+    on_finish: Option<F>,
 }
 
-impl<'a, W: Write> AutoFinishEncoder<'a, W> {
-    fn new<F>(encoder: Encoder<'a, W>, on_finish: F) -> Self
-    where
-        F: 'static + FnMut(io::Result<W>),
-    {
+impl<'a, W: Write, F> AutoFinishEncoder<'a, W, F>
+where
+    F: FnMut(io::Result<W>) + 'static,
```

Comment:
I don't think `'static` is still required, now that `F` is part of `Self` itself.